plugin/audit_log/audit_log.c

```diff
@@ -411,40 +450,48 @@ size_t audit_log_general_record(char *buf, size_t buflen,
                      "\"%s\",\"%s\",\"%s\",\"%s\",\"%lu\",%d,\"%s\",\"%s\","
                      "\"%s\",\"%s\",\"%s\"\n" };
 
-  return my_snprintf(buf, buflen,
-                     format_string[audit_log_format],
-                     name,
-                     make_record_id(id_str, sizeof(id_str)),
-                     make_timestamp(timestamp, sizeof(timestamp), t),
-                     event->general_sql_command.str,
-                     event->general_thread_id,
-                     status,
-                     escape_string(event->general_query,
-                                   event->general_query_length,
-                                   query, sizeof(query), NULL),
-                     escape_string(event->general_user,
-                                   event->general_user_length,
-                                   endptr, endtmp - endptr, &endptr),
-                     escape_string(event->general_host.str,
-                                   event->general_host.length,
-                                   endptr, endtmp - endptr, &endptr),
-                     escape_string(event->general_external_user.str,
-                                   event->general_external_user.length,
-                                   endptr, endtmp - endptr, &endptr),
-                     escape_string(event->general_ip.str,
-                                   event->general_ip.length,
-                                   endptr, endtmp - endptr, &endptr));
+  query= escape_string(event->general_query, event->general_query_length,
+                       endptr, endbuf - endptr, &endptr, &full_outlen);
+  user= escape_string(event->general_user, event->general_user_length,
+                      endptr, endbuf - endptr, &endptr, &full_outlen);
+  host= escape_string(event->general_host.str, event->general_host.length,
+                      endptr, endbuf - endptr, &endptr, &full_outlen);
+  external_user= escape_string(event->general_external_user.str,
+                               event->general_external_user.length,
+                               endptr, endbuf - endptr, &endptr, &full_outlen);
+  ip= escape_string(event->general_ip.str, event->general_ip.length,
+                    endptr, endbuf - endptr, &endptr, &full_outlen);
+
+  if (full_outlen > buflen / 2 - strlen(format_string[audit_log_format]) -
+                    MAX_RECORD_ID_SIZE - MAX_TIMESTAMP_SIZE)
+  {
+    return full_outlen * 2 + strlen(format_string[audit_log_format]) +
+           MAX_RECORD_ID_SIZE - MAX_TIMESTAMP_SIZE;
```

Comment:
Both condition and return statement need some adjustments. Here is better revision:
```c++
  size_t buflen_estimated= full_outlen * 2 +
                           strlen(format_string[audit_log_format]) +
                           strlen(name) +
                           strlen(event->general_sql_command.str) +
                           20 + /* general_thread_id */
                           20 + /* status */
                           MAX_RECORD_ID_SIZE + MAX_TIMESTAMP_SIZE;
  if (buflen_estimated > buflen)
  {
    return buflen_estimated;
  }
```